File: src/cli/version.test.ts
import { execFileSync } from "node:child_process";
import TempDir from "../TempDir.ts";
import { readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { test } from "node:test";

import { expect } from "expect";

const here = import.meta.dirname;
const tsx = join(here, "..", "..", "node_modules", ".bin", "tsx");
const cli = join(here, "main.ts");

const HAS_GIT = (() => {
  try {
    execFileSync("git", ["--version"], { stdio: "ignore" });
    return true;
  } catch {
    return false;
  }
})();

test(
  "cappu version bumps cappu.json (comments kept) and tags at a git root",
  { skip: !HAS_GIT },
  () => {
    using dir = TempDir.create("cappu-version-");
    const git = (...args: string[]): void => {
      execFileSync("git", args, { cwd: dir.path, stdio: "ignore" });
    };
    git("init");
    git("config", "user.email", "test@example.com");
    git("config", "user.name", "Test");
    writeFileSync(
      join(dir.path, "cappu.json"),
      '{\n  // my project\n  "groupId": "com.example",\n  "artifactId": "lib",\n  "version": "1.2.3"\n}\n',
    );
    git("add", ".");
    git("commit", "-m", "init");

    execFileSync(tsx, [cli, "version", "minor"], {
      cwd: dir.path,
      stdio: ["ignore", "pipe", "pipe"],
    });

    const after = readFileSync(join(dir.path, "cappu.json"), "utf8");
    expect(after).toContain('"version": "1.3.0"');
    expect(after).toContain("// my project"); // comment preserved
    const tags = execFileSync("git", ["tag"], { cwd: dir.path, encoding: "utf8" });
    expect(tags).toContain("v1.3.0");
    // the bump was committed as its own commit
    const subject = execFileSync("git", ["log", "-1", "--pretty=%s"], {
      cwd: dir.path,
      encoding: "utf8",
    });
    expect(subject.trim()).toBe("v1.3.0");
  },
);

test("cappu version still bumps cappu.json outside a git repo", { skip: !HAS_GIT }, () => {
  using dir = TempDir.create("cappu-version-");
  writeFileSync(join(dir.path, "cappu.json"), '{\n  "version": "0.9.0"\n}\n');
  execFileSync(tsx, [cli, "version", "patch"], {
    cwd: dir.path,
    stdio: ["ignore", "pipe", "pipe"],
  });
  expect(readFileSync(join(dir.path, "cappu.json"), "utf8")).toContain('"version": "0.9.1"');
});

File: togo/internal/cli/version_test.go
package cli

import (
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"testing"

	"github.com/nikeee/cappu/internal/config"
)

// Port of src/cli/version.test.ts. The TS test spawns the CLI binary; here we
// drive RunVersion directly (the dispatch wiring is covered by main).

func hasGit() bool {
	return exec.Command("git", "--version").Run() == nil
}

func loadDir(t *testing.T, dir string) *config.Config {
	t.Helper()
	cfg, err := config.Load("", dir)
	if err != nil {
		t.Fatalf("load config: %v", err)
	}
	return cfg
}

func TestVersionBumpsAndTagsAtGitRoot(t *testing.T) {
	if !hasGit() {
		t.Skip("git not available")
	}
	dir := t.TempDir()
	git := func(args ...string) {
		cmd := exec.Command("git", args...)
		cmd.Dir = dir
		if err := cmd.Run(); err != nil {
			t.Fatalf("git %v: %v", args, err)
		}
	}
	git("init")
	git("config", "user.email", "test@example.com")
	git("config", "user.name", "Test")
	cfgPath := filepath.Join(dir, "cappu.json")
	if err := os.WriteFile(cfgPath, []byte("{\n  // my project\n  \"groupId\": \"com.example\",\n  \"artifactId\": \"lib\",\n  \"version\": \"1.2.3\"\n}\n"), 0o644); err != nil {
		t.Fatal(err)
	}
	git("add", ".")
	git("commit", "-m", "init")

	if code := RunVersion("minor", "", loadDir(t, dir)); code != 0 {
		t.Fatalf("RunVersion exit = %d", code)
	}

	after, _ := os.ReadFile(cfgPath)
	if !strings.Contains(string(after), `"version": "1.3.0"`) {
		t.Errorf("version not bumped:\n%s", after)
	}
	if !strings.Contains(string(after), "// my project") {
		t.Errorf("comment not preserved:\n%s", after)
	}
	tags, _ := exec.Command("git", "-C", dir, "tag").Output()
	if !strings.Contains(string(tags), "v1.3.0") {
		t.Errorf("tag v1.3.0 missing, got %q", tags)
	}
	subject, _ := exec.Command("git", "-C", dir, "log", "-1", "--pretty=%s").Output()
	if strings.TrimSpace(string(subject)) != "v1.3.0" {
		t.Errorf("commit subject = %q, want v1.3.0", strings.TrimSpace(string(subject)))
	}
}

func TestVersionBumpsOutsideGitRepo(t *testing.T) {
	dir := t.TempDir()
	cfgPath := filepath.Join(dir, "cappu.json")
	if err := os.WriteFile(cfgPath, []byte("{\n  \"version\": \"0.9.0\"\n}\n"), 0o644); err != nil {
		t.Fatal(err)
	}
	if code := RunVersion("patch", "", loadDir(t, dir)); code != 0 {
		t.Fatalf("RunVersion exit = %d", code)
	}
	after, _ := os.ReadFile(cfgPath)
	if !strings.Contains(string(after), `"version": "0.9.1"`) {
		t.Errorf("version not bumped:\n%s", after)
	}
}
